### DISCUSSION TITLE
Argo CD notifications - get notified over failed repositories

### DISCUSSION THREAD

**Discussion by m8-t** [argoproj/argo-cd#16467] [Q&A] [answered]
Hello,

I am trying to setup Argo notifications. Here is my use case for clearance:

We use Argo CD as a multitentant GitOps tool. This workflow leads to some dependencies on external git/helm repositories. Because this Argo CD instance is highly dynamic and syncing/deployments fail regurlarly because of the developers doing developer things ;).
I do not want get a notification on app X that failed because of some Helm errors. Instead I want to monitor the health of the connected repositories and want to get notified when a repository is in failed state. There are some cases where the repositories go into an failure state, like a personal access token expired. I am not sure if there are kinda proper log entries to alert, thats why I thought about argo notifications with a custom trigger. 

With argocd cli and jq you can do something like this:

`argocd repo list -ojson | jq '.[] | select(.connectionState.status=="Failed") | "\(.repo)"'`

This gives you all repositories with status "Failed".

I wanted to build a  custom trigger from that info, so something like this:

```
trigger.repo-failed: |
  - description: Repo is failing
    send:
    - repo-failed
    when: repo.connectionState.status == 'Failed'
```

Anyways this doesn't work, the notification controller logs an error:

`time="2023-11-28T15:17:12Z" level=error msg="failed to execute when condition: cannot fetch status from <nil> (1:22)\n | repo.connectionState.status == 'Failed'\n | .....................^"`

I am not even sure if I can build custom triggers? Is there some kind of documentation on which fields from objects I can add to a trigger?

Any hint or is there another neat way to monitor the health of repos?

thanks

**Comment by mubarak-j** [comment, marked as answer]
I believe notification controller only watches custom resources of type 'application', therefore it can't track the repo status i.e the controller will always see `repo.connectionState.status == nil`

**Reply by m8-t** [reply]
Thanks for pointing it out.

**Reply by koushyk** [reply]
but u can use repo.GetCommitMetadata(app.status.sync.revision).Author in message templating. I've tried to use repo.GetCommitMetadata(app.status.sync.revision).Author in triggers to hide some notifications, but I failed
